File: python/day24.py
from functools import reduce


def parse_line(line):
    pos = 0
    while pos < len(line):
        if line[pos] in ['e', 'w']:
            yield line[pos]
            pos += 1
        elif line[pos:pos + 2] in ['se', 'sw', 'nw', 'ne']:
            yield line[pos:pos + 2]
            pos += 2
        else:
            return


def eval_step(pos, step):
    x, y = pos
    if len(step) == 1:
        x = x + 1 if step == 'e' else x - 1
    else:
        vdir, hdir = step[0], step[1]
        if y % 2 == 0:
            if hdir == 'w':
                x -= 1
        else:
            if hdir == 'e':
                x += 1
        y = y + 1 if vdir == 's' else y - 1
    return x, y


def get_neighbor_coord(pos):
    return (eval_step(pos, s) for s in
            ['e', 'w', 'se', 'sw', 'nw', 'ne'])


def eval_path(steps):
    return reduce(eval_step, steps, (0, 0))


def eval_paths(paths):
    res = set()
    for path in paths:
        pos = eval_path(path)
        if pos in res:
            res.remove(pos)
        else:
            res.add(pos)
    return res


def step_day(tiles):
    res = set()
    minx = min(tiles, key=lambda p: p[0])[0]
    maxx = max(tiles, key=lambda p: p[0])[0]
    miny = min(tiles, key=lambda p: p[1])[1]
    maxy = max(tiles, key=lambda p: p[1])[1]

    for x in range(minx - 1, maxx + 2):
        for y in range(miny - 1, maxy + 2):
            t = x, y
            num_neighbors = sum(
                npos in tiles for npos in get_neighbor_coord(t))
            if t in tiles:
                if (num_neighbors == 0 or num_neighbors > 2):
                    # flip to white
                    pass
                else:
                    res.add(t)
            elif num_neighbors == 2:
                res.add(t)
    return res


def solution():
    paths = [list(parse_line(line))
             for line in open('../data/24.txt', 'r').readlines()]

    tiles = eval_paths(paths)
    print(f'Answer 1: {len(tiles)}')

    for i in range(100):
        tiles = step_day(tiles)
    print(f'Answer 2: {len(tiles)}')

File: rust/src/day24.rs
use crate::common;
use std::collections::HashSet;

#[derive(Debug)]
enum HexDir {
  E,
  W,
  NW,
  SW,
  NE,
  SE,
}
type HexCoord = (i64, i64);

fn parse_path(line: &str) -> Vec<HexDir> {
  let mut it = line.chars();
  let mut res = Vec::new();
  loop {
    let dir = match it.next() {
      Some('e') => HexDir::E,
      Some('w') => HexDir::W,
      Some(x) => match (x, it.next()) {
        ('n', Some('w')) => HexDir::NW,
        ('n', Some('e')) => HexDir::NE,
        ('s', Some('w')) => HexDir::SW,
        ('s', Some('e')) => HexDir::SE,
        _ => return res,
      },
      _ => return res,
    };
    res.push(dir);
  }
}

fn eval_step((x, y): &HexCoord, step: &HexDir) -> HexCoord {
  let dy = match step {
    HexDir::SW | HexDir::SE => 1,
    HexDir::NW | HexDir::NE => -1,
    _ => 0,
  };
  let dx = match step {
    HexDir::E => 1,
    HexDir::W => -1,
    HexDir::SW | HexDir::NW if y % 2 == 0 => -1,
    HexDir::SE | HexDir::NE if y % 2 != 0 => 1,
    _ => 0,
  };
  (x + dx, y + dy)
}

fn eval_path(steps: &Vec<HexDir>) -> HexCoord {
  steps.iter().fold((0, 0), |pos, step| eval_step(&pos, step))
}

fn eval_paths(paths: &Vec<Vec<HexDir>>) -> HashSet<HexCoord> {
  let mut res = HashSet::new();
  for path in paths {
    let pos = eval_path(path);
    if res.contains(&pos) {
      res.remove(&pos);
    } else {
      res.insert(pos);
    }
  }
  res
}

fn step_day(tiles: &HashSet<HexCoord>) -> HashSet<HexCoord> {
  let minx = tiles.iter().map(|t| t.0).min().unwrap();
  let maxx = tiles.iter().map(|t| t.0).max().unwrap();
  let miny = tiles.iter().map(|t| t.1).min().unwrap();
  let maxy = tiles.iter().map(|t| t.1).max().unwrap();

  let mut res = HashSet::new();
  for x in (minx - 1)..=(maxx + 1) {
    for y in (miny - 1)..=(maxy + 1) {
      let pos = (x, y);
      let num_neighbors: u32 = [
        HexDir::E,
        HexDir::W,
        HexDir::NW,
        HexDir::SW,
        HexDir::NE,
        HexDir::SE,
      ]
      .iter()
      .map(|dir| tiles.contains(&eval_step(&pos, dir)) as u32)
      .sum();
      if tiles.contains(&pos) {
        if num_neighbors == 0 || num_neighbors > 2 {
        } else {
          res.insert(pos);
        }
      } else if num_neighbors == 2 {
        res.insert(pos);
      }
    }
  }
  res
}

pub(crate) fn solution() {
  let paths: Vec<Vec<HexDir>> = common::read_lines(&common::data_file(24))
    .unwrap()
    .iter()
    .map(|s| parse_path(s))
    .collect();

  let mut tiles = eval_paths(&paths);
  println!("Answer 1: {}", tiles.len());

  for _ in 0..100 {
    tiles = step_day(&tiles);
  }
  println!("Answer 2: {}", tiles.len());
}
